L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,471 @@
+
+  // Watches the updates of root certificates with name |root_cert_name| (set by
+  // the user via set_root_cert_name).  If used in TLS credentials, setting this
+  // field is optional for both the client side and the server side.  If this is
+  // not set on the client side, we will use the root certificates stored in the
+  // default system location, since client side must provide root certificates
+  // in TLS (no matter single-side TLS or mutual TLS).  If this is not set on
+  // the server side, we will not watch any root certificate updates, and assume
+  // no root certificates is needed for the server (in the one-side TLS
+  // scenario, the server is not required to provide root certificates). We
+  // don't support default root certs on server side.
+  void watch_root_certificates();
+
+  // Sets the name of root certificates being watched, if |watch_root_certificates| is
+  // called. If not set, an empty string will be used as the name. For simple
+  // use cases this is not needed, but a certificate provider could be
+  // implemented that watches many sets of certificates. This allows the user to
+  // differentiate between different root certificates.
+  // 
+  // @param root_cert_name the name of root certificates being set.
+  void set_root_cert_name(absl::string_view root_cert_name);
+
+  // Watches the updates of identity key-certificate pairs with name
+  // |identity_cert_name|. If used in TLS credentials, it is required to be set
+  // on the server side, and optional for the client side(in the one-side
+  // TLS scenario, the client is not required to provide identity certificates).
+  void watch_identity_key_cert_pairs();
+
+  // Sets the name of identity key-certificate pairs being watched, if
+  // |watch_identity_key_cert_pairs| is called. If not set, an empty string will
+  // be used as the name.
+  //
+  // @param identity_cert_name the name of identity key-certificate pairs being set.
+  void set_identity_cert_name(absl::string_view identity_cert_name);
+
+  // WARNING: EXPERT USE ONLY. MISUSE CAN LEAD TO SIGNIFICANT SECURITY DEGRADATION.
+  // 
+  // Sets the TLS session key logging file path. If not set, TLS
+  // session key logging is disabled. Note that this should be used only for
+  // debugging purposes. It should never be used in a production environment
+  // due to security concerns - anyone who can obtain the (logged) session key
+  // can decrypt all traffic on a connection.
+  // 
+  // @param tls_session_key_log_file_path: Path where TLS session keys should
+  // be logged.
+  void set_tls_session_key_log_file_path_dangerous(
+      absl::string_view tls_session_key_log_file_path);
+
+  // Sets the certificate verifier. The certificate verifier performs checks on
+  // the peer certificate chain after the chain has been (cryptographically)
+  // verified to chain up to a trusted root.
+  // If unset, this will default to the `HostNameCertificateVerifier` detailed
```

Comment:
That's a good question - I think it depends on other defaults, and if we are doing MTLS vs. TLS? I'll double check, but I don't believe this is called unless it is configured to do MTLS.

Reply:
I'm not sure that's correct. For example, we want it to default to the HostnameVerifier on the client-side even when we are doing (normal) TLS only.

Reply:
I meant on client side it would default to `HostNameCertificateVerifier`, but it wouldn't on the server side if we are only doing TLS, not mTLS?
This relates to that other comment, I need to draw out the links with the OpenSSL stack to see what of the custom verification is happening if we aren't doing verification

Reply:
The default (e.g. that SSL credentials and most TLS libraries have) should be:
- The client always does the hostname check.
- The server never does the hostname check, regardless of whether a client cert was requested or not.

Indeed, there is no meaningful "hostname check" that a server can perform on a client cert (e.g. what string is the server looking for in the client cert?).

Reply:
Understood, added in the comment that on client-side it defaults to the `HostNameVerifier` and on the server-side it defaults to nullptr